src/Cache.zig

```diff
@@ -762,18 +762,22 @@ pub const Manifest = struct {
 
     fn downgradeToSharedLock(self: *Manifest) !void {
         if (!self.have_exclusive_lock) return;
-        const manifest_file = self.manifest_file.?;
-        try manifest_file.downgradeLock();
+        if (std.process.can_spawn or !builtin.single_threaded) { // Some targets (WASI) do not support flock
+            const manifest_file = self.manifest_file.?;
+            try manifest_file.downgradeLock();
+        }
         self.have_exclusive_lock = false;
     }
 
     fn upgradeToExclusiveLock(self: *Manifest) !void {
         if (self.have_exclusive_lock) return;
-        const manifest_file = self.manifest_file.?;
-        // Here we intentionally have a period where the lock is released, in case there are
-        // other processes holding a shared lock.
-        manifest_file.unlock();
-        try manifest_file.lock(.Exclusive);
+        if (std.process.can_spawn or !builtin.single_threaded) { // Some targets (WASI) do not support flock
```

Comment:
This doesn't seem like it's checking the correct condition. Shouldn't it be something like "if the target supports file locking"?

Also I'm pretty sure file locking can be implemented in user space when it is not provided by the OS layer.

Reply:
Just to explain the thinking: This check is intending to verify that we won't have multiple processes or threads contending for the manifest file, in which case `flock` can be safely elided.

The user-space implementations that I know of are a wrapper around locks provided by `fcntl`, which are also unavailable on WASI. AFAIK, there's no way on WASI to implement file locking yet (both child process spawning and file locking are on the roadmap for consideration, though)

Reply:
It's entirely possible to use a shell to execute two different processes that are each single threaded and cannot spawn subprocesses but they each try to access the same file from the file system, right?

Reply:
Yeah, it is - Sounds like it will be best to constrain this workaround to WASI-only, and add a note referencing the open wasi-filesystem issue where file locking is being considered